### PR TITLE
Add ThreadPoolExecutor to workflowrunner

### LINKED ISSUE #802: Prevent leaking threads in workflow runner
workflow_runner needs someway of managing threads, perhaps with context manager. In order for our test environment to be sane, we need all spawned threads to be manageable.  Wrokflow_runner.run spawns a thread as a demon an exits. Spawning a thread as a demon should not be the only safeguard for it terminate in due time.

### PR DESCRIPTION
**Pull request by xjules**
**Issue**
Resolves #802 

**Approach**
Add ThreadPoolExecutor as an attribute to workflowrunner. 
Exception is provided workflowrunner.exception()

### CODE REVIEW

**Comment by xjules** [comment]
test this please!

File: python/res/job_queue/workflow_runner.py [hunk 1]
```diff
@@ -2,6 +2,20 @@
...
+    def __init__(self, target, name):
+        super(WorkflowRunnerThread, self).__init__(target=target, name=name)
+        self.setDaemon(True)
+        self._exception = None
+        # in python2.7 it's called __target therefore we need to inject this too
+        self._target = target
+
+    def run(self):
+        try:
+            if self._target:
```
Comment (markusdregi) [inline python/res/job_queue/workflow_runner.py]: Should this not be done by `super` instead? I.e. `super(WorkflowRunnerThread, self).run(self)`?

File: python/res/job_queue/workflow_runner.py [hunk 2]
```diff
@@ -2,6 +2,20 @@
 from res.job_queue import Workflow
 from res.util.substitution_list import SubstitutionList
 
+class WorkflowRunnerThread(Thread):
+    def __init__(self, target, name):
+        super(WorkflowRunnerThread, self).__init__(target=target, name=name)
+        self.setDaemon(True)
+        self._exception = None
+        # in python2.7 it's called __target therefore we need to inject this too
```
Comment (markusdregi) [inline python/res/job_queue/workflow_runner.py]: I don't understand this comment.

**Comment by markusdregi** [comment]
I still don't see how this PR is resolving the given issue. It solves the problem in the tests, but that is more by changing the test, then adopting the code at large (which I find a bit scary).

I'll have try to have a closer look at the usage of the WorkflowRunner and come back to you with something more concrete ;)

File: python/res/job_queue/workflow_runner.py [hunk 3]
```diff
@@ -41,9 +54,18 @@ def isCancelled(self):
...
+
+    @property
+    def runner_thread(self):
+        return self.__workflow_thread
 
     def wait(self):
-        self.__workflow.wait()
+        if self.runner_thread:
+            self.runner_thread.join()
+        print('DBG wait finnished with following exception', self.runner_thread._exception)
```
Comment (oysteoh) [inline python/res/job_queue/workflow_runner.py]: what happens if **not** self.runner_thread ? from what i understand your print will fail when calling _exception on it?

File: python/res/job_queue/workflow_runner.py (hunk 2 above)
Comment (xjules) [reply python/res/job_queue/workflow_runner.py]: The difference of Python2.7 vs Python3.3+ is `thread.__target` vs `thread._target`; therefore we set an attribute `thread._target` (which is tackled by __init__ for Python3.3+)

File: python/res/job_queue/workflow_runner.py (hunk 3 above)
Comment (xjules) [reply python/res/job_queue/workflow_runner.py]: Good point, should be inside the if statement

File: python/res/job_queue/workflow_runner.py (hunk 1 above)
Comment (xjules) [reply python/res/job_queue/workflow_runner.py]: Let's discuss this - need to know an opinion on this one

**Review by ManInFez — APPROVED** [review]
This looks good to me. It fixes the potential leaking of threads during testing. We should look into how Ert is using this and protect the execution in a similar manner.

**Review by markusdregi — CHANGES_REQUESTED** [review]
I would like to have another discussion regarding this PR before we merge it when Julius is back in the office :)

File: python/res/job_queue/workflow_runner.py [hunk 4]
```diff
@@ -20,12 +20,20 @@ def __init__(self, workflow, ert=None, context=None):
...
+
+    def __exit__(self, type, value, traceback):
+        self.wait()
 
     def run(self):
-        workflow_thread = Thread(name="ert_gui_workflow_thread")
-        workflow_thread.setDaemon(True)
-        workflow_thread.run = self.__runWorkflow
-        workflow_thread.start()
+        #should we check if it's already running?
```
Comment (markusdregi) [inline python/res/job_queue/workflow_runner.py]: If that is intended usage then please go ahead ;)

File: python/res/job_queue/workflow_runner.py [hunk 5]
```diff
@@ -20,12 +20,20 @@ def __init__(self, workflow, ert=None, context=None):
 
         self.__context = context
         self.__workflow_result = None
+        self.__workflow_executor = ThreadPoolExecutor(max_workers=1)
```
Comment (markusdregi) [inline python/res/job_queue/workflow_runner.py]: Double under scored prefix is reserved for Python. Please use single underscore instead 🙏

**Review by markusdregi — COMMENTED** [review]
Beautiful 👏 Can you evaluate the usage in _ERT_ and make an issue for changing the usage of `WorkflowRunner` in _ERT_ if that should be done?

File: python/res/job_queue/workflow_runner.py [hunk 6]
```diff
@@ -1,7 +1,7 @@
-from threading import Thread
 from res.job_queue import Workflow
 from res.util.substitution_list import SubstitutionList
-
+from concurrent.futures import ThreadPoolExecutor
+from concurrent import futures
```
Comment (markusdregi) [inline python/res/job_queue/workflow_runner.py:3]: Do we need both of these imports?

**Comment by markusdregi** [comment]
Do we need to update the requirements file and the Komodo dependencies of libres?

**Comment by ertomatic** [comment]
Can one of the admins verify this patch?

File: python/res/job_queue/workflow_runner.py (hunk 6 above)
Comment (xjules) [reply python/res/job_queue/workflow_runner.py:3]: No, will be gone ... soon :)

**Comment by xjules** [comment]
> Do we need to update the requirements file and the Komodo dependencies of libres?

No, we shouldn't.

File: python/res/job_queue/workflow_runner.py (hunk 5 above)
Comment (xjules) [reply python/res/job_queue/workflow_runner.py]: Ok, shall I change the existing ones too?

File: python/res/job_queue/workflow_runner.py (hunk 4 above)
Comment (xjules) [reply python/res/job_queue/workflow_runner.py]: ok

File: python/res/job_queue/workflow_runner.py (hunk 5 above)
Comment (xjules) [reply python/res/job_queue/workflow_runner.py]: A new issue spawned for existing ones #861

**Comment by markusdregi** [comment]
> > Do we need to update the requirements file and the Komodo dependencies of libres?
> 
> No, we shouldn't.

Are you sure? I don't think `futures` comes with Python 2.7...

**Comment by xjules** [comment]
> > > Do we need to update the requirements file and the Komodo dependencies of libres?
> > 
> > 
> > No, we shouldn't.
> 
> Are you sure? I don't think `futures` comes with Python 2.7...

~~That agrees, but it's included in libecl ... but can make it explicit in libres too.~~ That relates to `future` no `futures`
'futures` included in libres too now. The issue https://github.com/agronholm/pythonfutures/issues/29 only relates to spawning new processes, where in our case we spawn only new threads within the same process, so we are good to go!

File: python/res/job_queue/workflow_runner.py (hunk 4 above)
Comment (markusdregi) [reply python/res/job_queue/workflow_runner.py]: You should remove the comment and make a decision as to whether you want to check this or not ;) I would lean towards the later

Comment (xjules) [reply python/res/job_queue/workflow_runner.py]: Comment removed!

File: python/res/job_queue/workflow_runner.py [hunk 7]
```diff
@@ -20,12 +19,18 @@ def __init__(self, workflow, ert=None, context=None):
...
+
+    def __exit__(self, type, value, traceback):
+        self.wait()
 
     def run(self):
-        workflow_thread = Thread(name="ert_gui_workflow_thread")
-        workflow_thread.setDaemon(True)
-        workflow_thread.run = self.__runWorkflow
-        workflow_thread.start()
+        self._workflow_job = self._workflow_executor.submit(self.__runWorkflow)
```
Comment (markusdregi) [inline python/res/job_queue/workflow_runner.py]: Are we not leaking a thread now if we call run twice and then cancel arbitrary many times? If so, I would say that we should raise an Exception if `run` is called twice and it is clear misuse of the API.

Comment (xjules) [reply python/res/job_queue/workflow_runner.py]: The thread should not leak as it is the part of ThreadPool, i.e., just the job submitted twice - of-course it shouldn't be done for different reasons, so yes we can raise an exception here.

File: python/res/job_queue/workflow_runner.py [hunk 8]
```diff
@@ -20,12 +19,22 @@ def __init__(self, workflow, ert=None, context=None):
...
+    def __exit__(self, type, value, traceback):
+        self.wait()
 
     def run(self):
-        workflow_thread = Thread(name="ert_gui_workflow_thread")
-        workflow_thread.setDaemon(True)
-        workflow_thread.run = self.__runWorkflow
-        workflow_thread.start()
+        if self.isRunning():
+            raise Exception('An instance of workflow is already running!')
```
Comment (markusdregi) [inline python/res/job_queue/workflow_runner.py]: Never raw exceptions, perhaps an `AssertionError`?

Comment (xjules) [reply python/res/job_queue/workflow_runner.py]: ok. Was browsing around in libres and  found raw Exception still ....

**Review by markusdregi — APPROVED** [review]
Good job 👍 

Can you create a new tag on master listing that a new dependency is added after this is merged and add this to the dependency list for master and the new tag in komodo-releases?

**Pull request merged**